### PR TITLE
Add MTL data classes

### PR DESCRIPTION
**Pull request by bhancock8**
Port over from MeTaL/mmtl and Emmental

**Test plan**
- Port over and update existing tests for Dataset, DataLoader, and utility functions

### CODE REVIEW

File: snorkel/mtl/data.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
+import logging
+from collections import defaultdict
+
+import torch
+from torch.utils.data import DataLoader, Dataset
+
+from snorkel.mtl.utils import list_to_tensor
```
Comment (henryre) [inline snorkel/mtl/data.py]: nit: `from .utils import ...`

File: test/mtl/test_mtl_utils.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
+import unittest
```
Comment (henryre) [inline test/mtl/test_mtl_utils.py]: nit: rename `test_utils.py`

Comment (bhancock8) [reply test/mtl/test_mtl_utils.py]: I agree, but fun fact: tox won't let us have two tests with the same name:
```
import file mismatch:
imported module 'test_utils' has this __file__ attribute:
  /Users/braden/repos/snorkel/test/model/test_utils.py
which is not the same as the test file we want to collect:
  /Users/braden/repos/snorkel/test/mtl/test_utils.py
HINT: remove __pycache__ / .pyc files and/or use a unique basename for your test file modules
```

Comment (bhancock8) [reply test/mtl/test_mtl_utils.py]: It's actually a pytest issue:
https://github.com/pytest-dev/pytest/issues/529

**Pull request merged**